### ISSUE TITLE
[subset] pop_pack() option to not hash & share table

### ISSUE THREAD

**Issue by blueshade7** [harfbuzz/harfbuzz#2164]
As outlined in #1981, I am working on a task to modernize CFF/CFF2 subsetter using _subset2 without calculating table offsets in first pass, then actually serialize in second pass, as previously done with _subset.

Serializing INDEX table in a single pass is even possible by serializing all data first while calculating the minimum offSize required, then pop_pack() data so they move to tail of the serializer buffer, followed by serializing the INDEX header along with offsets using offSize calculated above.

Problem:

If I serialize sections of a contiguous table by calling pop_pack() piece-wise, some may be found in packed_map and actually not serialized, as the result a corrupted table is generated.

A possible workaround is to add a boolean flag to pop_pack() to not look it up in packed_map, so that it is surely serialized. This option may be also helpful to skip hashing a huge singleton sub-table unlikely shared, as discussed in above issue.

Ideas?

**Comment by blueshade7** [comment]
I came to think I am trying to use a facility (object packing) for something it was not intended for (swapping order of sub-sections of a table), so I withdraw this issue. I will fix my problem with INDEX table locally somehow.

An option to not hash an unlikely shared table may be worth to have though.

**Comment by blueshade7** [comment]
Let me reopen this. Struggling in some failed attempts, beginning to realize "not_to_map" pop_pack() option may be the easiest workaround to serialize FDArray in a single pass.

**Comment by garretrieger** [comment]
Seems reasonable to me.

**Comment by behdad** [comment]
I agree the not-sharing option is useful in a couple other places as well.  I'm happy to have that.  The objects should not be inserted in the sharing pool either in that case.  And yes, that sounds perfect solution for INDEX.

**Comment by behdad** [comment]
I'm reopening this to reconsider and solve differently.  Seems like being able to pack objects that will be packed "immediate after" current object is useful in quite a few places.  Even your INDEX case can be addressed this way.  INDEX is interesting because the offset size is also only known at runtime.  I want to come up with a generic-enough solution to even address that.

So I'm thinking that instead of "dont-share" we want "pack_immediate" or "pack_right"after" mode.  "pack_immediate" name might be a bit misleading.  I'm leaning on "pack_after".  One would be able to add offsets to these or even not and the packer will pack that object right after.  Multiple "pack_after" objects will be packed in the order they are added.

Now.  Initially I was thinking that the "afterness/immediateness" is a property of the `add_link`.  But I'm thinking it would be interesting to allow packing objects "after" even if no link is needed.  Basically allowing one to construct current object discontiguously.

As part of this I'm thinking about adding a ref-count to objects and implement garbage-collection.  That would be needed later anyway when we try to resolve offset overflows by duplicating later objects more closely.  The gc allows not keeping the later copy around if it's otherwise unused..

Am still thinking how to do this best.  It bothers me that typically a stack of items would be packed in reverse and then need to be reverted.  Am trying to think of how to do that best.  Packing them *not* in reverse order would require starting a new sub-serializer somewhere in the buffer since we need an additional "head" to keep track of those.  Am trying to avoid sub-serializer design.
